File: scripts/roster.ts
// Offline roster dump — the single source of truth for "which companies do the
// scrapers walk", read from the TypeScript rather than regexed out of it.
//
// The listed roster (auJobsTargets.ts) declares its entries as object literals,
// so a regex over the source could see them. The Top-150 private roster does
// NOT: TOP_PRIVATE_TARGETS is built with `RAW.map(buildPrivate)`, so its ids and
// names exist only once the module has run. Every scraper that regexed the
// source therefore walked 205 companies and silently skipped 150 — no error, no
// warning, just a smaller roster than anyone thought.
//
// Importing the modules is the fix: it is exactly what the cron worker composes
// into AU_JOBS_TARGETS, so the scrapers and the worker cannot disagree.
//
// Writes a JSON array of {id, name, sector, group, cities} on stdout.
// Run: bun run scripts/roster.ts
import { AU_JOBS_TARGETS } from "../src/employsi/data/auJobsTargets";
import { TOP_PRIVATE_TARGETS } from "../src/employsi/data/topPrivateCompanies";
import { UNIVERSITY_TARGETS } from "../src/employsi/data/universityTargets";
import { CITY_ROSTERS } from "../src/employsi/data/cityRosters";
import { rosterId } from "../src/employsi/data/rosters";

// City-roster companies (the exchange-listed names plotted on each global hub).
// Included so the international scrapers can walk them too — Naukri needs the
// Mumbai and Bengaluru names, which exist only here and not in the AU rosters.
const CITY_TARGETS = Object.entries(CITY_ROSTERS).flatMap(([city, r]) =>
  r.companies.map(([ticker, name, group]) => ({
    id: rosterId(city, ticker),
    name,
    sector: group,
    group,
    cities: [city],
  })),
);

// The city rosters are OPT-IN (--with-cities). Folding them in by default would
// silently take every existing driver — Indeed, LinkedIn, JobsDB, Jora — from
// 355 companies to ~995 and change what those daily runs mean. Only the
// scrapers that actually cover those markets ask for them.
const withCities = process.argv.includes("--with-cities");
// KEEP IN STEP WITH workers/jobs-cron/index.ts, which composes the same three
// AU files for the Adzuna pull. Two independent assemblies of one roster: a
// category added there and not here is a set of companies the Worker pulls that
// no Python driver walks, and neither side errors. check-feed-coverage.py
// compares the two for exactly that reason.
const AU_ALL = [...AU_JOBS_TARGETS, ...TOP_PRIVATE_TARGETS, ...UNIVERSITY_TARGETS];
const all = withCities ? [...AU_ALL, ...CITY_TARGETS] : AU_ALL;
const seen = new Set<string>();
const out = all.filter((t) => {
  if (!t.id || seen.has(t.id)) return false;
  seen.add(t.id);
  return true;
});
process.stdout.write(JSON.stringify(out));

File: scripts/roster.py
#!/usr/bin/env python3
"""
The company roster, shared by every scraper driver.

Reads it out of the TypeScript by RUNNING the modules (scripts/roster.ts via
bun) rather than regexing the source. That distinction matters: the listed
roster declares object literals a regex can see, but the Top-150 private roster
is built with `RAW.map(buildPrivate)`, so its ids and names only exist once the
module has run. Drivers that regexed the source walked 205 companies and
silently skipped 150.

Which is why there is NO regex fallback here. A fallback would reintroduce the
exact failure this replaces — a short roster that looks like a successful run.
If bun is missing or the dump fails, this raises, and the caller stops.
"""
from __future__ import annotations
import json
import os
import subprocess

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)


class RosterError(RuntimeError):
    pass


def load_roster(only: set[str] | None = None, limit: int | None = None,
                with_cities: bool = False) -> list[dict]:
    """Every company the scrapers walk: {id, name, sector, group, cities}.

    `only` filters to a set of ids; `limit` truncates (both after the full list
    is loaded, so a --limit run is a prefix of the real roster rather than a
    different roster).

    `with_cities` adds the exchange-listed companies plotted on each global hub
    (cityRosters.ts) — about 640 more names, including the Mumbai and Bengaluru
    ones. It is OFF by default so the Australian drivers keep walking exactly
    the 355 companies they always have; only scrapers that cover those markets
    ask for them.
    """
    try:
        p = subprocess.run(
            ["bun", "run", os.path.join(HERE, "roster.ts")]
            + (["--with-cities"] if with_cities else []),
            capture_output=True,
            timeout=120,
            cwd=ROOT,
        )
    except FileNotFoundError as e:
        raise RosterError(
            "bun is required to read the company roster "
            "(the private roster is generated, so it cannot be read from source). "
            "Install bun: https://bun.sh"
        ) from e
    if p.returncode != 0:
        raise RosterError(f"roster dump failed: {p.stderr.decode()[:300]}")
    try:
        rows = json.loads(p.stdout.decode())
    except Exception as e:
        raise RosterError(f"roster dump was not JSON: {p.stdout[:200]!r}") from e
    if not isinstance(rows, list) or not rows:
        raise RosterError("roster dump was empty")
    if only:
        rows = [r for r in rows if r.get("id") in only]
    if limit is not None:
        rows = rows[:limit]
    return rows


if __name__ == "__main__":
    r = load_roster()
    print(f"{len(r)} companies")
    for x in r[:3]:
        print(" ", x["id"], "|", x["name"], "|", x.get("sector"))
